In [1]:
class Neuron:
    def __init__(self, i, w):
        if len(w)!=len(i):
            print("inputs and weight does not match")
            return
        self.inputs = i
        self.weights = w
        self.output = 0
        self.previouses = None
        self.linear_combiner()

    def linear_combiner(self):
        """
        inputs : list
        gets input or set of inputs from the network
        """
        for i in range(len(self.inputs)):
            self.output += self.inputs[i]*self.weights[i]
        return self.activation()

    def activation(self):
        """
        Uses ReLU - piecewise linear function that will output the input directly 
        if it is positive, otherwise, it will output zero        
        """
        if self.output < 0:
            self.output = 0
        return self.output

    def add_previous(self, neuron):
        if self.previouses is None:
            self.previouses = [neuron]
        else:
            self.previouses.append(neuron)
        
        self.inputs.append(neuron.inputs)
        self.weights.append(neuron.weights)


In [2]:
def neuron_table(inputs, weights):
    """
    inputs : list - array of numbers from previous layer or the input
    weights : list - array of numbers from the link of the inputs

    Displays the tally of computation of each node
    """
    print("Input\tWeight\tProduct\tReLU\tCurrent Sum")
    sum_of_prod = 0
    for i in range(len(inputs)):
        product = inputs[i]*weights[i]
        sum_of_prod += product if product > 0 else 0
        print(inputs[i],'\t',weights[i],'\t',product,end='\t')
        print(product if product > 0 else 0,"\t",sum_of_prod)
    print("SUM of products:",sum_of_prod)

In [3]:
inputs = [5,6,7,8,9]
weights_1 = [1,3,5,7,9]

In [4]:
n1 = Neuron(inputs,weights_1)

neuron_table(inputs,weights_1)
print('\nNeuron 1:',n1.output)

Input	Weight	Product	ReLU	Current Sum
5 	 1 	 5	5 	 5
6 	 3 	 18	18 	 23
7 	 5 	 35	35 	 58
8 	 7 	 56	56 	 114
9 	 9 	 81	81 	 195
SUM of products: 195

Neuron 1: 195


The product of the first input with its weight is 5. The second is 18; third is 35; 4th is 56; and the 5th is 81. Upon using ReLU, their value remained the same

In [5]:
n2 = Neuron(inputs,weights_1)

neuron_table(inputs,weights_1)
print('\nNeuron 2:',n2.output)

Input	Weight	Product	ReLU	Current Sum
5 	 1 	 5	5 	 5
6 	 3 	 18	18 	 23
7 	 5 	 35	35 	 58
8 	 7 	 56	56 	 114
9 	 9 	 81	81 	 195
SUM of products: 195

Neuron 2: 195


In [6]:
n3 = Neuron(inputs,weights_1)

neuron_table(inputs,weights_1)
print('\nNeuron 3:',n3.output)

Input	Weight	Product	ReLU	Current Sum
5 	 1 	 5	5 	 5
6 	 3 	 18	18 	 23
7 	 5 	 35	35 	 58
8 	 7 	 56	56 	 114
9 	 9 	 81	81 	 195
SUM of products: 195

Neuron 3: 195


ReLU works


In [7]:
from random import randint

#Random weights for the output neuron
output_weights = [randint(-1,5) for _ in range(3)]
print(output_weights)

[0, 2, 1]


In [8]:
output_n = Neuron([n1.output, n2.output, n3.output],output_weights)

neuron_table([n1.output, n2.output, n3.output],output_weights)
print('\nOutput neuron:',output_n.output)

Input	Weight	Product	ReLU	Current Sum
195 	 0 	 0	0 	 0
195 	 2 	 390	390 	 390
195 	 1 	 195	195 	 585
SUM of products: 585

Output neuron: 585


New Perceptron Task
Create a network with feed forward with multiple neurons

In [13]:
import numpy as np
class Neuron:
    def __init__(self, inputs, weights, activation):
        self.inputs = inputs
        self.weights = weights
        self.output = 0
        self.act = activation
        self.linear_combiner()

    def linear_combiner(self):
        """
        inputs : list
        gets input or set of inputs from the network
        """
        for i in range(len(self.inputs)):
            self.output += self.inputs[i]*self.weights[i]
        return self.activation()

    def activation(self):
        """
        Uses ReLU - piecewise linear function that will output the input directly 
        if it is positive, otherwise, it will output zero 
        Uses sigmoid - takes any real value as input and outputs values in the range of 0 to 1       
        """
        if self.act == 'ReLU':
            if self.output < 0:
                self.output = 0
        elif self.act == 'sigmoid':
            self.output = 1 / (1 + np.exp(-self.output))
        else:
            print("Invalid activation function")
            return None
        return self.output

class Layer:
    def __init__(self, num_neurons: int, inputs: list, weights: list, activation='sigmoid'):
        self.neurons = [Neuron(inputs, weights, activation) for _ in range(num_neurons)]
    
    def get_output(self):
        outputs = []
        for neuron in self.neurons:
            outputs.append(neuron.output)
        return outputs

- Redefined the neuron to have more activation function
- Created a layer class that has the collection of neurons (They have identical activation function and neighboring layers)

In [384]:
inputs = [5,6,7,8,9]
weights_1 = [1,3,5,7,9]
output_weights = [np.random.randint(-1,4) for _ in range(5)]

input_layer = Layer(5, inputs, weights_1, activation="ReLU")
output_layer = Layer(1, input_layer.get_output(), output_weights, activation="sigmoid")

print("Input\tWeight\tInput Layer\tWeight\tOutput Layer")
for i in range(5):
    print(inputs[i],end='\t')
    print(weights_1[i],end='\t')
    print(input_layer.get_output()[i],end='\t\t')
    print(output_weights[i],end='\t')
    if i==0:
        print(output_layer.get_output()[0],end="")
    print("")

Input	Weight	Input Layer	Weight	Output Layer
5	1	195		2	1.0
6	3	195		1	
7	5	195		3	
8	7	195		2	
9	9	195		1	


The first three columns shown are identical to the displayed output of the three prototype neurons above. All neurons in the input layer have identical values (195). The random weight assigned with random randint were paired with the 5 neuron values. Since the sum of all the the produce of the paired input layer and weight is huge, the sigmoid activation function return 1.